### ISSUE TITLE
TextSpan with WidgetSpan Fails

### ISSUE THREAD

**Issue by junioregis** [flutter/flutter#49883]
Hello Flutter Dev Team! :)

When I use **TextSpan** with **WidgetSpan**, the **TextPainter** fails on calling **tp.layout()**.

          var span = TextSpan(
            children: [
                WidgetSpan(child: Icon(Icons.access_time)), 
                TextSpan(text: "Clock")
            ],
          );

          var tp = TextPainter(
            maxLines: 10,
            textAlign: TextAlign.left,
            textDirection: TextDirection.ltr,
            text: span,
          );

          tp.layout(
            maxWidth: size.maxWidth,
          );

**Flutter (Channel stable, v1.12.13+hotfix.7, on Linux, locale en_US.UTF-8)**

```
════════ Exception caught by widgets library ═══════════════════════════════════════════════════════
The following assertion was thrown building LayoutBuilder:
'package:flutter/src/widgets/widget_span.dart': Failed assertion: line 103 pos 12: 'dimensions != null': is not true.

When the exception was thrown, this was the stack: 
#2      WidgetSpan.build (package:flutter/src/widgets/widget_span.dart:103:12)
#3      TextSpan.build (package:flutter/src/painting/text_span.dart:211:15)
#4      TextPainter.layout (package:flutter/src/painting/text_painter.dart:544:13)
```

**Comment by iapicca** [comment]
Hi @junioregis 
does this [code](https://dartpad.dev/30bd7b3cc2e7a5c9c6c56c7ddd5b0365) represent the issue you are describing?
Thank you

**Comment by junioregis** [comment]
> Hi @junioregis
> does this [code](https://dartpad.dev/30bd7b3cc2e7a5c9c6c56c7ddd5b0365) represent the issue you are describing?
> Thank you

**Hi @iapicca! Replace this code:**

```
text: const TextSpan(
      text: 'Hello, world.',
      style:  TextStyle(
        color: Colors.black,
        fontSize: 30,
      ),
),
```

**To this:**

```
children: [
  TextSpan(text: "Opss..."),
  WidgetSpan(child: Icon(Icons.timelapse)),
],
```

**Comment by lukaspili** [comment]
Same issue when using `InlineSpan` as a child of `TextSpan`.

**Comment by jason-simmons** [comment]
Call `TextPainter.setPlaceholderDimensions` to provide dimensions for the `WidgetSpan` before calling `TextPainter.layout`.

Or use the `RichText` widget, which internally lays out the embedded widgets and calls `setPlaceholderDimensions`.

**Comment by lukaspili** [comment]
It's unclear how to use `TextPainter.setPlaceholderDimensions` with the following `WidgetSpan`:

```dart
final widgetSpan = WidgetSpan(
    alignment: PlaceholderAlignment.middle,
    child: CustomPaint(
      painter: SomeCustomPainterThatDrawsABackgroundColor(),
      child: Text('Blah'),
    ),
);

final painter = TextPainter(
  text: TextSpan(
    text: 'Foo',
    children: [widgetSpan],
  ),
);
painter.setPlaceholderDimensions(?);
painter.layout();
```

The placeholder dimension's size should be equal to the wrapped Text('Blah') size, right?

The desired outcome is to measure the height of the whole text before rendering. So I don't think I can use `RichText` directly.

**Comment by medyas** [comment]
I am having a similar problem with WidgetSpan, as it is not correctly positioned in the RichText. Here is a screen of what am getting as output:
![Capture](https://user-images.githubusercontent.com/13151636/73581438-81604600-4489-11ea-80fa-3896bf0d25d4.PNG)
 i created a dartpad file for test : [`DartPad`](https://dartpad.dev/b91e6fc5df31eaf98ac52c8254536f51) but the WidgetSpan will throw an error #42086
As you can see, some of the WidgetSpan are positioned on top of the text.
Code: 
```
RichText(
      textDirection: TextDirection.rtl,
      softWrap: true,
     textAlign: TextAlign.justify,
      textScaleFactor: 2.0,
      text: TextSpan(
        style: const TextStyle(color: Colors.black),
        children: widgetList,
      ),
    );
```
It is the same problem even when using `TextAlign.center`

Flutter Doctor:
`[√] Flutter (Channel stable, v1.12.13+hotfix.7, on Microsoft Windows [Version 10.0.17763.914], locale en-US)`

**Comment by tobiww** [comment]
Possibly related to [51306]( https://github.com/flutter/flutter/issues/51306) ?

**Comment by pedromassangocode** [comment]
I was able to reproduce this issue on latest master channel.

A simple usage of the class `WidgetSpan` throws an exception. The cause of the issue is just:

`WidgetSpan(child: Icon(Icons.timelapse))`

<details><summary>Logs</summary>

```
Performing hot restart...
Syncing files to device iPhone 11 Pro Max...
Restarted application in 1,082ms.

════════ Exception caught by rendering library ═════════════════════════════════════════════════════
The following assertion was thrown during paint():
'package:flutter/src/widgets/widget_span.dart': Failed assertion: line 103 pos 12: 'dimensions != null': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md

The relevant error-causing widget was: 
  CustomPaint file:///Users/pedromassango/dev/Projects/flutter_example/lib/main.dart:17:12
When the exception was thrown, this was the stack: 
#2      WidgetSpan.build (package:flutter/src/widgets/widget_span.dart:103:12)
#3      TextSpan.build (package:flutter/src/painting/text_span.dart:212:15)
#4      TextPainter.layout (package:flutter/src/painting/text_painter.dart:569:14)
#5      MyPainter.paint (package:flutter_example/main.dart:39:7)
#6      RenderCustomPaint._paintWithPainter (package:flutter/src/rendering/custom_paint.dart:531:13)
...
The following RenderObject was being processed when the exception was fired: RenderCustomPaint#a428c relayoutBoundary=up2
...  parentData: offset=Offset(57.0, 298.0) (can use size)
...  constraints: BoxConstraints(0.0<=w<=414.0, 0.0<=h<=896.0)
...  size: Size(300.0, 300.0)
RenderObject: RenderCustomPaint#a428c relayoutBoundary=up2
  parentData: offset=Offset(57.0, 298.0) (can use size)
  constraints: BoxConstraints(0.0<=w<=414.0, 0.0<=h<=896.0)
  size: Size(300.0, 300.0)
════════════════════════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by rendering library ═════════════════════════════════════════════════════
The following assertion was thrown during paint():
'package:flutter/src/widgets/widget_span.dart': Failed assertion: line 103 pos 12: 'dimensions != null': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md

The relevant error-causing widget was: 
  CustomPaint file:///Users/pedromassango/dev/Projects/flutter_example/lib/main.dart:17:12
When the exception was thrown, this was the stack: 
#2      WidgetSpan.build (package:flutter/src/widgets/widget_span.dart:103:12)
#3      TextSpan.build (package:flutter/src/painting/text_span.dart:212:15)
#4      TextPainter.layout (package:flutter/src/painting/text_painter.dart:569:14)
#5      MyPainter.paint (package:flutter_example/main.dart:39:7)
#6      RenderCustomPaint._paintWithPainter (package:flutter/src/rendering/custom_paint.dart:531:13)
...
The following RenderObject was being processed when the exception was fired: RenderCustomPaint#a428c relayoutBoundary=up2
...  parentData: offset=Offset(57.0, 298.0) (can use size)
...  constraints: BoxConstraints(0.0<=w<=414.0, 0.0<=h<=896.0)
...  size: Size(300.0, 300.0)
RenderObject: RenderCustomPaint#a428c relayoutBoundary=up2
  parentData: offset=Offset(57.0, 298.0) (can use size)
  constraints: BoxConstraints(0.0<=w<=414.0, 0.0<=h<=896.0)
  size: Size(300.0, 300.0)
════════════════════════════════════════════════════════════════════════════════════════════════════

```
</details>

<details><summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.356, on Mac OS X 10.15.6 19G2021 x86_64, locale en)
    • Flutter version 1.22.0-10.0.pre.356 at /Users/pedromassango/dev/SDKs/flutter_master
    • Framework revision c412fd9967 (4 hours ago), 2020-09-24 22:47:03 -0700
    • Engine revision 435db234e4
    • Dart version 2.11.0 (build 2.11.0-161.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0, Build version 12A7209
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Community Edition (version 2020.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 49.0.4
    • Dart plugin version 202.7319.5

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.1

[✓] Connected device (4 available)
    • iPhone 11 Pro Max (mobile) • 7A82F29E-E49E-4B3C-8609-1570361DBC17 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)            • macos                                • darwin-x64     • Mac OS X 10.15.6 19G2021 x86_64
    • Web Server (web)           • web-server                           • web-javascript • Flutter Tools
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 85.0.4183.121

! Doctor found issues in 1 category.
Process finished with exit code 0
```
</details>

**Comment by josefrvaldes** [comment]
Is this going to be fixed at some point? I'm having this issue now trying to use a WidgetSpan inside a TextSpan.

Or at least, could you clarify how to workaround the issue by using setPlaceholderDimensions?

Thanks!

**Comment by PassionPenguin** [comment]
> I was able to reproduce this issue on latest master channel.
> 
> A simple usage of the class `WidgetSpan` throws an exception. The cause of the issue is just:
> 
> `WidgetSpan(child: Icon(Icons.timelapse))`
> 
> Logs
> flutter doctor -v

well it's a big problem and lasts for a long decade. do you have any solution to this problem, for now?

**Comment by kodaira313ayahoo** [comment]
> EDIT: Actually solved it by measuring individually each `Text` wrapped by the `WidgetSpan` and passing these sizes to the placeholder dimensions.

@lukaspili 
May I ask you of how to do it? I don't know how to do both "how to wrap" and "how to pass the sizes 
 to the placeholder dimensions". Especially I want to know the latter. Thank you.

**Comment by danagbemava-nc** [comment]
It looks like this hasn't been documented yet 
https://master-api.flutter.dev/flutter/painting/TextPainter/setPlaceholderDimensions.html

**Comment by felipecastrosales** [comment]
hey, any workaround?

**Comment by jobehi** [comment]
**Hello, here is a workaround :** 

tl;dr you can find the two workarounds in this [gist](https://gist.github.com/jobehi/5e07636a2c1b55dbd9ec6f02a617ad19)

In order for it to layout, the `TextPainter` needs to know the placeholder dimensions so that it leaves spaces in between the `TextSpans`. 
For that, the `setPlaceholderDimensions` should be called with these dimensions. 

**How to get the dimensions ?**

First of all, you need to get the `RenderObject` of the `TextSpan` / `InlineSpan` widget.
To do so, the simplest way is to assign a key to your widget and then get the corresponding `RenderObject` from its `context`

```

/// example

 final richTextWithWidgetSpanKey = GlobalKey();

class RichTextWithWidgetSpanCase extends StatelessWidget {
  const RichTextWithWidgetSpanCase({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return const Text.rich(
      TextSpan(
        text: 'Case 3 - Rich text with Widget spans ',
        children: <InlineSpan>[
          WidgetSpan(
            child: Icon(Icons.star, size: 16),
          ),
          TextSpan(
            text: 'bold',
            style: TextStyle(
              fontWeight: FontWeight.bold,
            ),
          ),
          WidgetSpan(
            child: Icon(Icons.star, size: 16),
          ),
          TextSpan(text: ' world!'),
        ],
      ),
    );
  }
}

RichTextWithWidgetSpanCase(key: richTextWithWidgetSpanKey)

```
Then 

```
RenderParagraph? _getRenderParagraphFromKey(GlobalKey key) {
    final keyContext = key.currentContext;
    if (keyContext == null) {
      return null;
    }
    final renderObject = keyContext.findRenderObject();
    if (renderObject is! RenderParagraph) {
      return null;
    }
    return renderObject;
  }
  
  
final renderParagraph = _getRenderParagraphFromKey(globalKey);
```

After getting the renderObject, you need to create the TextPainter normally 

```
 final renderParagraph = _getRenderParagraphFromKey(globalKey);
      if (renderParagraph == null) {
        return;
      }
      final text = renderParagraph.text;
      final textPainter = TextPainter(
        text: text,
        textDirection: TextDirection.ltr,
        textAlign: renderParagraph.textAlign,
        textScaler: renderParagraph.textScaler,
        maxLines: renderParagraph.maxLines,
        strutStyle: renderParagraph.strutStyle,
        textWidthBasis: renderParagraph.textWidthBasis,
        textHeightBehavior: renderParagraph.textHeightBehavior,
      );
```
and right before calling  `textPainter.setPlaceholderDimensions(placeholderDimensions)`  `textPainter.layout()`  you have two workaround

**Option 1 : The simple way - Use the protected method** : `layoutInlineChildren`

```
  List<PlaceholderDimensions>? _getPlaceHolderDimensions(
      RenderParagraph renderParagraph) {
    // this method is protected in the TextPainter class
    // ignore: invalid_use_of_protected_member
    return renderParagraph.layoutInlineChildren(
      renderParagraph.paintBounds.width,
      (RenderBox child, BoxConstraints constraints) =>
          Size(child.getMinIntrinsicWidth(double.infinity), 0.0),
      ChildLayoutHelper.getDryBaseline,
    );
  }
  
  // then you call this method to get the list and add it to the `textPainter` before layout 
  
  final placeholderDimensions = _getPlaceHolderDimensions(renderParagraph);

  textPainter.setPlaceholderDimensions(placeholderDimensions);

  textPainter.layout(minWidth: 0, maxWidth: renderParagraph.paintBounds.width);

```
A safer solution would be to deport the logic inside that method to your own code.

**Option 2 : reimplement the private internal methods your code** 

A simple naive solution would be to extend the RenderParagraph and copy the internal and private methods there

- `_layoutChild`
```
extension on RenderParagraph {

  /// call this method to get the PlaceholderDimensions
  List<PlaceholderDimensions>? getPlaceHolderDimensions() {
    return _layoutInlineChildren(
      paintBounds.width,
      (RenderBox child, BoxConstraints constraints) =>
          Size(child.getMinIntrinsicWidth(double.infinity), 0.0),
      ChildLayoutHelper.getDryBaseline,
    );
  }
}

  PlaceholderDimensions _layoutChild(
      RenderBox child,
      BoxConstraints childConstraints,
      ChildLayouter layoutChild,
      ChildBaselineGetter getBaseline) {
    final TextParentData parentData = child.parentData! as TextParentData;
    final PlaceholderSpan? span = parentData.span;
    assert(span != null);
    return span == null
        ? PlaceholderDimensions.empty
        : PlaceholderDimensions(
            size: layoutChild(child, childConstraints),
            alignment: span.alignment,
            baseline: span.baseline,
            baselineOffset: switch (span.alignment) {
              PlaceholderAlignment.aboveBaseline ||
              PlaceholderAlignment.belowBaseline ||
              PlaceholderAlignment.bottom ||
              PlaceholderAlignment.middle ||
              PlaceholderAlignment.top =>
                null,
              PlaceholderAlignment.baseline =>
                getBaseline(child, childConstraints, span.baseline!),
            },
          );
  }

  List<PlaceholderDimensions>? _layoutInlineChildren(double maxWidth,
      ChildLayouter layoutChild, ChildBaselineGetter getChildBaseline) {
    final BoxConstraints constraints = BoxConstraints(maxWidth: maxWidth);
    return <PlaceholderDimensions>[
      for (RenderBox? child = firstChild;
          child != null;
          child = childAfter(child))
        _layoutChild(child, constraints, layoutChild, getChildBaseline),
    ];
  }

  ```
Then you just call the method 

```
final placeholderDimensions = renderParagraph.getPlaceHolderDimensions();

textPainter.setPlaceholderDimensions(placeholderDimensions);

textPainter.layout(minWidth: 0, maxWidth: renderParagraph.paintBounds.width);

```

And voilà, you have a working painter. 

Note : While these workarounds provide a solution, they are somewhat involved. I plan to submit a pull request to make these methods more accessible and to propose a public API for retrieving placeholders. Since this would be my first contribution to the repository, it might take some time to finalize. In the meantime, I can develop a small package to encapsulate this functionality.

In the meantime, I can create a small package for that

@felipecastrosales @kodaira313ayahoo @PassionPenguin @josefrvaldes @pedromassangocode
